include/tvm/relay/optimizer.h

```diff
@@ -0,0 +1,341 @@
+/*!
+ * Copyright (c) 2019 by Contributors
+ * \file tvm/relay/optimizer.h
+ *
+ * \brief The optimizer manages a sequence of Relay-to-Relay transformation
+ * passes over a particlar unit of AST. The design is largely inspired from
+ * LLVM's pass manager.
+ */
+#ifndef TVM_RELAY_OPTIMIZER_H_
+#define TVM_RELAY_OPTIMIZER_H_
+
+#include <tvm/attrs.h>
+#include <tvm/ir.h>
+#include <tvm/packed_func_ext.h>
+#include <tvm/relay/error.h>
+#include <tvm/relay/expr.h>
+#include <tvm/relay/module.h>
+#include <tvm/relay/type.h>
+
+#include <string>
+#include <vector>
+
+namespace tvm {
+namespace relay {
+namespace optimize {
+
+/*! \brief An enumerator to represent the granularity of different passes. */
+enum PassKind : int {
+  kModuleKind = 1,
+  kFunctionKind = 2
+};
+
+class PassContext;
+
+/*
+ * \brief PassContextNode contains the information that a pass can rely on, such as
+ * the analysis result.
+ */
+class PassContextNode : public RelayNode {
+ public:
+  /*!
+   * \brief The error reporter used to notify users why an optimization fails.
+   */
+  ErrorReporter err_reporter_;
+
+  PassContextNode() = default;
+
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+  }
+
+  TVM_DLL static PassContext make();
+
+  static constexpr const char* _type_key = "relay.PassContext";
+  TVM_DECLARE_NODE_TYPE_INFO(PassContextNode, RelayNode);
+};
+
+class PassContext : public NodeRef {
+ public:
+  PassContext() = default;
+  explicit PassContext(NodePtr<tvm::Node> p) : NodeRef(p) {}
+
+  const PassContextNode* operator->() const {
+    return static_cast<PassContextNode*>(this->node_.get());
+  }
+
+  using ContainerType = PassContextNode;
+};
+
+// We use currying here. The Relay module is captured for optimizations. It
```

Comment:
Could you elaborate on `module is captured for optimizations` and why curry is needed?